### ISSUE TITLE
Authorized routine doesnt work on private dataset after routine update

### ISSUE THREAD

**Issue by marky7** [terraform-google-modules/terraform-google-bigquery#244]
### TL;DR

After creating authorized routine in BigQuery on private dataset, routine can query private table from private dataset. 
If routine is updated, routine lost access to private dataset after routine update. Its authorization is still present in dataset but not taken into account.

### Expected behavior

After routine update, the dataset authorizations should be still active.

### Observed behavior

I apply the documentation about [creating authorized_routine](https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/bigquery_dataset_access.html)

![PR_bq_authz drawio2](https://github.com/terraform-google-modules/terraform-google-bigquery/assets/11940761/433ba71f-bcdc-4094-8d20-190237656077)

After creating Bigquery resources :
- Public routine in public dataset 
- Private table in private dataset 
- Routine access of public routine to private dataset

The service account can query the public routine and get results from private dataset. 

`bq query --use_legacy_sql=false 'SELECT * FROM `public_dataset.get_items_by_timestamp_range`("2022-11-09 14:19:24.206000 UTC", "2023-11-09 14:20:45.270000 UTC");'
`
But if I update the routine, authorizations are invalidated and then I get the following error : 

BigQuery error in query operation: Error processing job 'terraform-[ID1]:bqjob_[ID2]': Access Denied: Table terraform-[ID1]:private_dataset.private_table: User does not have permission to query table
terraform-[ID1]:private_dataset.private_table, or perhaps it does not exist in location EU.

However if I display private dataset configurations, the routine authorization is still present :

```
bq show --format=prettyjson private_dataset  
```
```
    {
      "routine": {
        "datasetId": "private_dataset",
        "projectId": "terraform-[ID]",
        "routineId": "get_items_by_timestamp_range"
      }
    }
```

Workaround : 

When I update the routine, I also update the google_bigquery_dataset_access resource name. 
Terraform recreates the google_bigquery_dataset_access so it works.

### Terraform Configuration

```hcl
terraform {
  required_version = ">= 0.15.0"
  backend "gcs" {}
}

provider "google" {
  region  = "europe-west1"
  version = "~> 4.72.0"
  project = var.PROJECT_ID
}

provider "google-beta" {
  version = "~> 4.72.0"
}

data "google_client_config" "default" {}


...


##################################################################################
#                      BIGQUERY ROUTINES for End Users                           #
#                                                                                #
#                         ! DEVELOPER WARNING !                                  #
#                                                                                #
# 🚫 Modifying this ROUTINE will INVALIDATE its access to the private dataset.   #
#                                                                                #
# To prevent disruption, if you change any routine attribute :                   #
#                                                                                #
# 🛠️ Rename the 'google_bigquery_dataset_access' to regenerate it concurrently,  #
# appending the modification date/time.                                          #
#                                                                                #
# EXAMPLE: 'authorize_routine_get_items_by_timestamp_range_20230717_1730'      #
#                                                                                #
# This ensures unintended loss of dataset access.                                #
#                                                                                #
##################################################################################

resource "google_bigquery_routine" "get_items_by_timestamp_range" {
  dataset_id      = google_bigquery_dataset.public_dataset.dataset_id
  project         = var.project
  description     = <<-EOS
  Table valued function routine that filter on authorized data
  Usage :
  SELECT * FROM ${google_bigquery_dataset.public_dataset.dataset_id}.get_items_by_timestamp_range('2022-11-09 14:19:24.206000 UTC', '2023-11-09 14:20:45.270000 UTC');
  EOS
  routine_id      = "get_items_by_timestamp_range"
  routine_type    = "TABLE_VALUED_FUNCTION"
  language        = "SQL"
  definition_body = <<-EOS
  SELECT id, timestamp
  FROM `${var.project}.${google_bigquery_dataset.private_table.dataset_id}.${google_bigquery_table.private_table.table_id}`
  WHERE timestamp >= startDate
  AND timestamp <= endDate
  EOS
  arguments {
    name          = "startDate"
    argument_kind = "FIXED_TYPE"
    data_type     = jsonencode({ "typeKind" : "TIMESTAMP" })
  }
  arguments {
    name          = "endDate"
    argument_kind = "FIXED_TYPE"
    data_type     = jsonencode({ "typeKind" : "TIMESTAMP" })
  }

  return_table_type = jsonencode({ "columns" : [
    {
      "name" : "id",
      "type" : { "typeKind" : "STRING" }
    },
    {
      "name" : "timestamp",
      "type" : { "typeKind" : "TIMESTAMP" }
    }
  ] })
}

##################################################################################
#                       Routines Access to Dataset                               #
#                                                                                #
#                         ! DEVELOPER WARNING !                                  #
#                                                                                #
# 🛠️ On routine modifications, update 'google_bigquery_dataset_access' using the #
# specified naming convention. This applies to routine changes ONLY.             #
#                                                                                #
# EXAMPLE: 'authorize_routine_get_items_by_timestamp_range_20230717_1730'        #
#                                                                                #
# Adhering to this naming convention ensures dataset access is preserved after   #
# making changes to the routine.                                                 #
#                                                                                #
##################################################################################

resource "google_bigquery_dataset_access" "authorize_routine_get_items_by_timestamp_range_20230717_1737" {
  dataset_id = google_bigquery_dataset.private_table.dataset_id
  project    = var.project
  routine {
    project_id = var.project
    dataset_id = google_bigquery_dataset.public_dataset.dataset_id
    routine_id = google_bigquery_routine.get_items_by_timestamp_range.routine_id
  }
}
```

### Terraform Version

```
Terraform v1.5.5
on linux_amd64
```

### Additional information

Steps snippets for testing : 

#### GCloud

gcloud auth application-default login
gcloud config set [PROJECT]

#### TERRAFORM 

cd ~/terraform-bug-fix/blueprint
terraform init -var-file="../environments/dev.tfvars"
terraform refresh -var-file="../environments/dev.tfvars"
terraform plan -var-file="../environments/dev.tfvars"
terraform apply -var-file="../environments/dev.tfvars"

#### BIGQUERY

cd ~/terraform-bug-fix/sa
gcloud auth activate-service-account --key-file='./sa.json'
gcloud config set project [PROJECT]
bq ls --routines public_dataset
bq query --use_legacy_sql=false 'SELECT * FROM `public_dataset.get_items_by_timestamp_range`("2022-11-09 14:19:24.206000 UTC", "2023-11-09 14:20:45.270000 UTC");'

**Comment by marky7** [comment]
Hello Team,

I hope everyone is doing well. I wanted to touch base on the issue, which is still a bit of a challenge for us. I completely understand how busy things can get, but if you could spare a moment to revisit and possibly reopen this issue, it would mean a lot.

Warm regards

**Comment by LavishMishraApex** [comment]
Hi Team,

We use Table Functions as a tool for exposing data outside of a database to other projects with authorized routines. Can you please revisit this issue and check if it would be possible to work on it. Thanks in Advance!
